File: common/src/main/java/kg/delletenebre/yamus/api/response/Track.kt
package kg.delletenebre.yamus.api.response

import kotlinx.serialization.Serializable

@Serializable
data class Track(
        val id: String = "null",
        var playlistId: String = "",
        val albums: List<Album> = listOf(),
        val artists: List<Artist> = listOf(),
        val available: Boolean = false,
        val availableForPremiumUsers: Boolean = false,
        val availableFullWithoutPermission: Boolean = false,
        val contentWarning: String = "",
        var coverUri: String = "",
        val durationMs: Long = 0,
        val fileSize: Int = 0,
        val lyricsAvailable: Boolean = false,
        val major: Major = Major(),
        val normalization: Normalization = Normalization(),
        val ogImage: String = "",
        val previewDurationMs: Int = 0,
        val realId: String = "",
        val storageDir: String = "",
        val title: String = "",
        val type: String = "",
        val version: String = ""
) {
    companion object {
        const val DOWNLOAD_STATUS_PROGRESS = "progress"
        const val DOWNLOAD_STATUS_DOWNLOADED = "downloaded"
        const val DOWNLOAD_STATUS_ERROR = "error"
    }
    var downloadStatus = ""
    var downloadProgress = 0
    var playingState: String = ""

    @Serializable
    data class Major(
            val id: Int = 0,
            val name: String = ""
    )

    @Serializable
    data class Normalization(
            val gain: Double = 0.0,
            val peak: Int = 0
    )

    fun getTrackId(): String {
        var result = id
        if (albums.isNotEmpty()) {
            result += ":${albums[0].id}"
        }
        return result
    }

    fun getArtistName(): String {
        return if (artists.isNotEmpty()) {
            artists[0].name
        } else {
            ""
        }
    }
}

File: lib/api/models/track.dart
import 'package:audio_service/audio_service.dart';
import 'package:yamus/api/models.dart';

class Track {
  const Track({
    this.id = '',
    this.realId = '',
    this.title = '',
    this.version = '',
    this.available = false,
    this.durationMs = 0,
    this.normalization = const Normalization(),
    this.artists = const [],
    this.coverUri = '',
    this.contentWarning = '',
  });

  final String id;
  final String realId;
  final String title;
  final String version;
  final bool available;
  final int durationMs;
  final Normalization normalization;
  final List<Artist> artists;
  final String coverUri;
  final String contentWarning; // explicit
  // final List<Album> albums; // TODO parse album

  factory Track.fromJson(Map<String, dynamic> json) {
    final artists = List<Artist>.from(
      json['artists'].map((artist) => Artist.fromJson(artist))
    );

    return Track(
      id: json['id'],
      realId: json['realId'],
      title: json['title'],
      version: json['version'] ?? '',
      available: json['available'],
      durationMs: json['durationMs'],
      normalization: Normalization.fromJson(json['normalization']),
      artists: artists,
      coverUri: json['coverUri'],
      contentWarning: json['contentWarning'] ?? '',
    );
  }
}

extension TrackExtensions on Track {
  MediaItem asMediaItem() {
    return MediaItem(
      id: this.id,
      album: 'album',
      title: this.title,
      duration: Duration(milliseconds: this.durationMs),
      artist: this.artists.map((artist) => artist.name).join(', '),
    );
  }
  // ···
}